sycl/doc/extensions/proposed/sycl_ext_oneapi_device_image_content.asciidoc

```diff
@@ -0,0 +1,254 @@
+The format of this data is implementation-defined.
+See below for a description of this format on {dpcpp}.
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+std::span<std::byte> ext_oneapi_get_content_view() const;
+----
+!====
+
+Available on when the compiler is {cpp}20 or higher.
+
+_Returns:_ A view of the raw bytes for this device image.
+The data behind this view has the same lifetime as the `kernel_bundle` that
+contains this `device_image`.
+
+'''
+
+
+== Device image format for {dpcpp}
+
+This section is non-normative and applies only to the {dpcpp} implementation.
+The format of the data returned by `device_image::ext_oneapi_get_content` and
+`device_image::ext_oneapi_get_content_view` depends on the backend of the kernel
+bundle that contains the device image and also on the `State` of the device
+image.
+
+=== Format on Level Zero
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The format could change in the future, but it is
+  currently a SPIR-V module representing one or more kernels.
+  This SPIR-V module may be partially linked, with references to internal
+  library functions that are not defined in the SPIR-V module.
+
+* `bundle_state::object`: The same format as `bundle_state::input`.
+
+* `bundle_state::executable`: The device image content is native ISA for the
+  device, which can be passed to `zeModuleCreate` as `ZE_MODULE_FORMAT_NATIVE`
+  format.
+
+:ref1: ../proposed/sycl_ext_oneapi_free_function_kernels.asciidoc#level-zero-and-opencl-compatibility
+
+[_Note:_ The interface to kernels in the device image is not defined in the
+general case, which means there is no portable way to invoke kernels from a
+Level Zero module that is created from the raw device image content.
+However, see link:{ref1}[here] for a limited case where this portability is
+guaranteed.
+_{endnote}_]
+
+=== Format on OpenCL
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The format could change in the future, but it is
+  currently a SPIR-V module as described above for Level Zero.
+
+* `bundle_state::object`: The device image content is an unspecified format that
+  is created by calling `clCompileProgram`.
+
+* `bundle_state::executable`: The device image content is executable binary
+  device code representing one or more kernels, which can be passed to
+  `clCreateProgramWithBinary`.
+
+[_Note:_ The interface to kernels in the device image is not defined in the
+general case, which means there is no portable way to invoke kernels from a
+OpenCL `cl_program` object that is created from the raw device image content.
+However, see link:{ref1}[here] for a limited case where this portability is
+guaranteed.
+_{endnote}_]
+
+=== Format on CUDA
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The device image content is a PTX module representing
+  one or more kernels.
+
+* `bundle_state::object`: ???
```

Comment:
@AerialMantis: Can you help with this part?  What is the device image content for an `object` state bundle on CUDA?

Reply:
There is no meaningful state for the cuda and HIP backend, they each ignore transition to this state and jump to the last.

Reply:
That sounds the same as as with Level Zero.  Can we say:

> The same format as `bundle_state::input`.

Reply:
I decided to restrict this extension to only `bundle_state::executable`, so I think the issue here is now moot.